sql/core/src/main/scala/org/apache/spark/sql/execution/r/MapPartitionsRWrapper.scala

```diff
@@ -25,6 +25,21 @@ import org.apache.spark.sql.Row
 import org.apache.spark.sql.types.{BinaryType, StructField, StructType}
 
 /**
+ * A function wrapper that applies the given R function to each partition of each group.
+ */
+private[sql] case class MapGroupsRWrapper(
```

Comment:
repartition guarantees all items in a same group should be in a same partition, however, it does not group them. while in this PR, sort is required during shuffle in the physical operator:
```
  override def requiredChildDistribution: Seq[Distribution] =
    ClusteredDistribution(groupingAttributes) :: Nil

  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
    Seq(groupingAttributes.map(SortOrder(_, Ascending)))
```
so we can simply have groups: (key, iter)